hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/HoodieFileIndex.scala

```diff
@@ -233,8 +233,23 @@ case class HoodieFileIndex(spark: SparkSession,
       //    - Col-Stats Index is present
       //    - Record-level Index is present
       //    - List of predicates (filters) is present
+      val prunedPartitionFileNames: Set[String] = {
+        prunedPartitionsAndFileSlices
+          .flatMap {
+            case (_, fileSlices) => fileSlices
+          }
+          .flatMap { fileSlice =>
+            val baseFileOption = Option(fileSlice.getBaseFile.orElse(null))
+            val logFiles = if (includeLogFiles) {
+              fileSlice.getLogFiles.iterator().asScala.map(_.getFileName).toList
+            } else Nil
+            baseFileOption.map(_.getFileName).toList ++ logFiles
+          }
+          .toSet
+      }
+
       val candidateFilesNamesOpt: Option[Set[String]] =
-      lookupCandidateFilesInMetadataTable(dataFilters) match {
+      lookupCandidateFilesInMetadataTable(dataFilters, prunedPartitionFileNames) match {
```

Comment:
I mean we can put transformation inside `lookupCandidateFilesInMetadataTable`, after `!isMetadataTableEnabled || !isDataSkippingEnabled` is checked.